File: src/test/java/com/gisiona/demodevinia/infrastructure/adapter/persistence/UserRepositoryAdapterTest.java
package com.gisiona.demodevinia.infrastructure.adapter.persistence;

import com.gisiona.demodevinia.domain.model.User;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class UserRepositoryAdapterTest {

    @Mock
    private UserJpaRepository userJpaRepository;

    private UserRepositoryAdapter userRepositoryAdapter;

    @BeforeEach
    void setUp() {
        userRepositoryAdapter = new UserRepositoryAdapter(userJpaRepository);
    }

    @Test
    void shouldSaveUserSuccessfully() {
        LocalDateTime now = LocalDateTime.now();
        User user = User.create("João", "joao@example.com");
        UserEntity savedEntity = new UserEntity("João", "joao@example.com", now, now);
        savedEntity.setId(1L);

        when(userJpaRepository.save(any(UserEntity.class))).thenReturn(savedEntity);

        User result = userRepositoryAdapter.save(user);

        assertEquals(1L, result.getId());
        assertEquals("João", result.getName());
        assertEquals("joao@example.com", result.getEmail());
        verify(userJpaRepository).save(any(UserEntity.class));
    }

    @Test
    void shouldFindUserByIdSuccessfully() {
        Long id = 1L;
        LocalDateTime now = LocalDateTime.now();
        UserEntity entity = new UserEntity("João", "joao@example.com", now, now);
        entity.setId(id);

        when(userJpaRepository.findById(id)).thenReturn(Optional.of(entity));

        Optional<User> result = userRepositoryAdapter.findById(id);

        assertTrue(result.isPresent());
        assertEquals(id, result.get().getId());
        assertEquals("João", result.get().getName());
        assertEquals("joao@example.com", result.get().getEmail());
        verify(userJpaRepository).findById(id);
    }

    @Test
    void shouldReturnEmptyWhenUserNotFoundById() {
        Long id = 1L;

        when(userJpaRepository.findById(id)).thenReturn(Optional.empty());

        Optional<User> result = userRepositoryAdapter.findById(id);

        assertFalse(result.isPresent());
        verify(userJpaRepository).findById(id);
    }

    @Test
    void shouldFindUserByEmailSuccessfully() {
        String email = "joao@example.com";
        LocalDateTime now = LocalDateTime.now();
        UserEntity entity = new UserEntity("João", email, now, now);
        entity.setId(1L);

        when(userJpaRepository.findByEmail(email)).thenReturn(Optional.of(entity));

        Optional<User> result = userRepositoryAdapter.findByEmail(email);

        assertTrue(result.isPresent());
        assertEquals(1L, result.get().getId());
        assertEquals("João", result.get().getName());
        assertEquals(email, result.get().getEmail());
        verify(userJpaRepository).findByEmail(email);
    }

    @Test
    void shouldReturnEmptyWhenUserNotFoundByEmail() {
        String email = "joao@example.com";

        when(userJpaRepository.findByEmail(email)).thenReturn(Optional.empty());

        Optional<User> result = userRepositoryAdapter.findByEmail(email);

        assertFalse(result.isPresent());
        verify(userJpaRepository).findByEmail(email);
    }

    @Test
    void shouldFindAllUsersSuccessfully() {
        LocalDateTime now = LocalDateTime.now();
        UserEntity entity1 = new UserEntity("João", "joao@example.com", now, now);
        entity1.setId(1L);
        UserEntity entity2 = new UserEntity("Maria", "maria@example.com", now, now);
        entity2.setId(2L);
        List<UserEntity> entities = Arrays.asList(entity1, entity2);

        when(userJpaRepository.findAll()).thenReturn(entities);

        List<User> result = userRepositoryAdapter.findAll();

        assertEquals(2, result.size());
        assertEquals("João", result.get(0).getName());
        assertEquals("Maria", result.get(1).getName());
        verify(userJpaRepository).findAll();
    }

    @Test
    void shouldDeleteUserById() {
        Long id = 1L;

        userRepositoryAdapter.deleteById(id);

        verify(userJpaRepository).deleteById(id);
    }

    @Test
    void shouldCheckIfUserExistsByEmail() {
        String email = "joao@example.com";

        when(userJpaRepository.existsByEmail(email)).thenReturn(true);

        boolean result = userRepositoryAdapter.existsByEmail(email);

        assertTrue(result);
        verify(userJpaRepository).existsByEmail(email);
    }

    @Test
    void shouldReturnFalseWhenUserDoesNotExistByEmail() {
        String email = "joao@example.com";

        when(userJpaRepository.existsByEmail(email)).thenReturn(false);

        boolean result = userRepositoryAdapter.existsByEmail(email);

        assertFalse(result);
        verify(userJpaRepository).existsByEmail(email);
    }
}

File: src/test/kotlin/com/gisiona/demodevinia/infrastructure/persistence/adapter/UserRepositoryAdapterTest.kt
package com.gisiona.demodevinia.infrastructure.persistence.adapter

import com.gisiona.demodevinia.domain.model.User
import com.gisiona.demodevinia.infrastructure.persistence.entity.UserEntity
import com.gisiona.demodevinia.infrastructure.persistence.repository.JpaUserRepository
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Test
import java.util.*
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

class UserRepositoryAdapterTest {

    private val jpaUserRepository = mockk<JpaUserRepository>()
    private val userRepositoryAdapter = UserRepositoryAdapter(jpaUserRepository)

    @Test
    fun `should save user and return domain object`() {
        val user = User(name = "John Doe", email = "john@example.com")
        val userEntity = UserEntity(
            id = user.id,
            name = user.name,
            email = user.email,
            createdAt = user.createdAt,
            updatedAt = user.updatedAt
        )

        every { jpaUserRepository.save(any<UserEntity>()) } returns userEntity

        val result = userRepositoryAdapter.save(user)

        assertEquals(user.name, result.name)
        assertEquals(user.email, result.email)
        verify { jpaUserRepository.save(any<UserEntity>()) }
    }

    @Test
    fun `should find user by id and return domain object`() {
        val id = UUID.randomUUID()
        val userEntity = UserEntity(id = id, name = "John Doe", email = "john@example.com")

        every { jpaUserRepository.findById(id) } returns Optional.of(userEntity)

        val result = userRepositoryAdapter.findById(id)

        assertEquals(userEntity.name, result?.name)
        assertEquals(userEntity.email, result?.email)
        verify { jpaUserRepository.findById(id) }
    }

    @Test
    fun `should return null when user not found by id`() {
        val id = UUID.randomUUID()

        every { jpaUserRepository.findById(id) } returns Optional.empty()

        val result = userRepositoryAdapter.findById(id)

        assertNull(result)
        verify { jpaUserRepository.findById(id) }
    }

    @Test
    fun `should find user by email and return domain object`() {
        val email = "john@example.com"
        val userEntity = UserEntity(name = "John Doe", email = email)

        every { jpaUserRepository.findByEmail(email) } returns userEntity

        val result = userRepositoryAdapter.findByEmail(email)

        assertEquals(userEntity.name, result?.name)
        assertEquals(userEntity.email, result?.email)
        verify { jpaUserRepository.findByEmail(email) }
    }

    @Test
    fun `should return null when user not found by email`() {
        val email = "john@example.com"

        every { jpaUserRepository.findByEmail(email) } returns null

        val result = userRepositoryAdapter.findByEmail(email)

        assertNull(result)
        verify { jpaUserRepository.findByEmail(email) }
    }

    @Test
    fun `should delete user when exists`() {
        val id = UUID.randomUUID()

        every { jpaUserRepository.existsById(id) } returns true
        every { jpaUserRepository.deleteById(id) } returns Unit

        val result = userRepositoryAdapter.deleteById(id)

        assertTrue(result)
        verify { jpaUserRepository.existsById(id) }
        verify { jpaUserRepository.deleteById(id) }
    }
}
